app/models/deleted_object.rb

```diff
@@ -8,5 +8,76 @@ class DeletedObject < ApplicationRecord
     scope scoped_class.to_s.underscore.pluralize.to_sym, -> { where(object_type: scoped_class) }
   end
 
-  # TODO: cleanup regularly once the owner is destroyed too :) or this will grow uncontrollably
+  scope :owned_by_service, -> { where.has { owner_type == Service } }
+
+  scope :missing_owner, lambda {
+    associations = distinct(:owner_type).pluck(:owner_type).map(&:constantize).map do |association|
```

Comment:
I was giving a second thought on this. This is not thread-safe.

Say we have _N_ distinct types of owners in `deleted_objects`. When we invoke the scope, `distinct(:owner_type).pluck(:owner_type)` will set the  array of types based on which we'll loop and add _N_ `JOIN` clauses to the scope. In between these two steps (fetching the types and fetching the records) however, a _N+1_-th distinct type of owner can be inserted into `deleted_objects`. That specific ongoing call to scope will miss the records whose `owner_type` value is equal to new type just inserted.

Whilst `missing_owner` is only used for fetching stale records within the scope of a `PurgeStaleObjectsWorker` job (or other low risk operations), perhaps it's all right to miss a few records every once in a while. For more critical operations on the other hand, we may need to reconsider this approach.